framework/audio/engine/internal/audioplayer.cpp

```diff
@@ -239,33 +261,22 @@ bool AudioPlayer::hasTracks() const
     return m_tracks.size() > 0;
 }
 
-ValCh<PlayStatus> AudioPlayer::status() const
+void AudioPlayer::onMidiPlayContextChanged(const engine::Context& ctx)
 {
-    return m_status;
+    //LOGI() << ctx.dump();
+
+    if (ctx.hasVal(CtxKey::PlayTick)) {
+        uint32_t tick = ctx.get<uint32_t>(CtxKey::PlayTick);
+        if (tick != m_lastMidiPlayTick) {
+            m_lastMidiPlayTick = tick;
+            m_midiTickPlayed.send(tick);
+        }
+    }
 }
 
-void AudioPlayer::onPlayCallbackCalled()
+void AudioPlayer::onMidiStatusChanged(engine::IAudioEngine::Status status)
 {
-    //! NOTE For tests in development
-//    struct Time {
-//        QElapsedTimer time;
-//        int last = 0;
-//        Time() { time.start(); }
-//    };
-
-//    static Time time;
-
-//    static float lastPos = 0;
-
-//    if (m_status.val == PlayStatus::PLAYING) {
-//        float p = playbackPosition();
-//        float deltaPos = p - lastPos;
-//        lastPos = p;
-
-//        int cur = time.time.elapsed();
-//        int delta = cur - time.last;
-//        time.last = cur;
-
-//        LOGI() << "[onPlayCallbackCalled] pos: " << p << ", delta pos: " << deltaPos << ", timer: " << delta;
-//    }
+    if (status == engine::IAudioEngine::Status::Stoped) {
```

Comment:
Why don't we have the same subscription for "pause" for example? Isn't it necessary?

Reply:
The player has its own status and it controls it.
But there is only a problem with the stop status.
The fact is that when the sound source ends, then the audioengine destroyed it (equivalent to the action of the stops).
It is about this situation that the player needs to know, that the audioengine has destroyed (stopped) the source, because it has ended.

There will be more complex logic in the future. When the player controls more than one midi handler, but has several tracks at the same time